File: src/flavor-go/pkg/psp/format_2025/locking.go
// SPDX-FileCopyrightText: Copyright (c) 2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package format_2025

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync/atomic"
	"time"

	"log/slog"
)

// Global flag for lock acquisition status
var lockAcquired int32

var isProcessRunningFn = IsProcessRunning
var lockFprintfFn = fmt.Fprintf
var openLockFileFn = func(path string, flag int, perm os.FileMode) (*os.File, error) {
	return os.OpenFile(path, flag, perm)
}

// LockInfo represents lock file information
type LockInfo struct {
	PID int `json:"pid"`
}

// TryAcquireLock attempts to acquire an exclusive lock for cache extraction
// Returns true if lock was acquired, false if cache is already being extracted
func TryAcquireLock(paths *WorkenvPaths, logger *slog.Logger) (bool, error) {
	// Create instance/extract directory if it doesn't exist
	extractDir := paths.Extract()
	if err := os.MkdirAll(extractDir, os.FileMode(DirPerms)); err != nil {
		logger.Debug("Failed to create extract directory", "error", err)
	}

	lockPath := paths.LockFile()
	pid := os.Getpid()

	// Check for stale lock first
	if _, err := os.Stat(lockPath); err == nil {
		logger.Debug("🔍 Lock file exists, checking if it's stale...")

		// Try to read the PID from the lock file
		if data, err := os.ReadFile(lockPath); err == nil {
			contents := strings.TrimSpace(string(data))
			if oldPid, err := strconv.Atoi(contents); err == nil {
				if !isProcessRunningFn(oldPid) {
					logger.Info("🧹 Removing stale lock from dead process", "pid", oldPid)
					_ = os.Remove(lockPath)
				} else {
					logger.Debug("🔒 Lock held by active process", "pid", oldPid)
					return false, nil
				}
			} else {
				// Invalid PID in lock file, remove it
				logger.Info("🧹 Removing invalid lock file (couldn't parse PID)")
				_ = os.Remove(lockPath)
			}
		} else {
			// Can't read lock file, try to remove it
			logger.Info("🧹 Removing unreadable lock file")
			_ = os.Remove(lockPath)
		}
	}

	// Try to create lock file exclusively
	file, err := openLockFileFn(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, FilePerms)
	if err != nil {
		if os.IsExist(err) {
			logger.Debug("🔒 Lock file exists, another process is extracting")
			return false, nil
		}
		return false, err
	}
	defer func() { _ = file.Close() }()

	// Write our PID to the lock file
	if _, err := lockFprintfFn(file, "%d\n", pid); err != nil {
		_ = os.Remove(lockPath)
		return false, err
	}

	logger.Debug("🔒 Acquired extraction lock", "pid", pid)
	atomic.StoreInt32(&lockAcquired, 1)
	return true, nil
}

// ReleaseLock releases the extraction lock
func ReleaseLock(paths *WorkenvPaths, logger *slog.Logger) {
	lockPath := paths.LockFile()
	if err := os.Remove(lockPath); err != nil {
		logger.Debug("⚠️ Failed to remove lock file", "error", err)
	} else {
		logger.Debug("🔓 Released extraction lock")
	}
	atomic.StoreInt32(&lockAcquired, 0)
}

// WaitForExtraction waits for another process to finish extraction
func WaitForExtraction(paths *WorkenvPaths, timeoutSecs int, logger *slog.Logger) error {
	lockPath := paths.LockFile()
	maxAttempts := timeoutSecs * 10 // Check every 100ms

	for attempt := 0; attempt < maxAttempts; attempt++ {
		if _, err := os.Stat(lockPath); os.IsNotExist(err) {
			logger.Debug("✅ Extraction lock released, cache should be ready")
			// Give a bit more time for files to be fully written
			time.Sleep(100 * time.Millisecond)
			return nil
		}

		if attempt%10 == 0 {
			logger.Debug("⏳ Waiting for extraction to complete...",
				"elapsed", fmt.Sprintf("%d/%ds", attempt/10, timeoutSecs))
		}

		time.Sleep(100 * time.Millisecond)
	}

	return fmt.Errorf("timeout waiting for cache extraction to complete")
}

// MarkExtractionComplete marks cache extraction as complete
func MarkExtractionComplete(paths *WorkenvPaths, logger *slog.Logger) error {
	extractDir := paths.Extract()
	if err := os.MkdirAll(extractDir, os.FileMode(DirPerms)); err != nil {
		return err
	}
	markerPath := paths.CompleteFile()
	file, err := os.Create(markerPath)
	if err != nil {
		return err
	}
	defer func() { _ = file.Close() }()

	if _, err := lockFprintfFn(file, "%d\n", os.Getpid()); err != nil {
		return err
	}
	logger.Debug("✅ Marked extraction as complete")
	return nil
}

// IsExtractionComplete checks if cache extraction is complete
func IsExtractionComplete(paths *WorkenvPaths) bool {
	_, err := os.Stat(paths.CompleteFile())
	return err == nil
}

// MarkExtractionIncomplete marks cache as incomplete (used during signal handling)
func MarkExtractionIncomplete(paths *WorkenvPaths, logger *slog.Logger) {
	extractDir := paths.Extract()
	if err := os.MkdirAll(extractDir, os.FileMode(DirPerms)); err != nil {
		logger.Warn("⚠️ Failed to create extract dir", "error", err)
	}
	// Remove the complete marker if it exists
	_ = os.Remove(paths.CompleteFile())
	logger.Debug("⚠️ Marked extraction as incomplete")
}

// IsLockAcquired checks if lock is currently acquired
func IsLockAcquired() bool {
	return atomic.LoadInt32(&lockAcquired) != 0
}

// CleanupStaleExtractions cleans up stale extraction directories from dead processes
func CleanupStaleExtractions(paths *WorkenvPaths, logger *slog.Logger) error {
	tmpDir := paths.Tmp()

	// If the directory doesn't exist, nothing to clean
	if _, err := os.Stat(tmpDir); os.IsNotExist(err) {
		return nil
	}

	// List all directories in tmp/
	entries, err := os.ReadDir(tmpDir)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		if entry.IsDir() {
			// Try to parse PID from directory name
			if pid, err := strconv.Atoi(entry.Name()); err == nil {
				// Check if process is still running
				if !isProcessRunningFn(pid) {
					staleDir := filepath.Join(tmpDir, entry.Name())
					logger.Info("🧹 Cleaning up stale extraction directory from dead process", "pid", pid)
					if err := os.RemoveAll(staleDir); err != nil {
						logger.Debug("⚠️ Failed to remove stale directory", "path", staleDir, "error", err)
					}
				}
			}
		}
	}

	return nil
}

File: src/flavor-rs/src/psp/format_2025/locking.rs
//! Lock file management for concurrent execution safety

use crate::exceptions::{FlavorError, Result};
use log::{debug, info};
use std::fs;
use std::io::Write;
use std::sync::atomic::{AtomicBool, Ordering};

use super::paths::WorkenvPaths;

/// Global flag for lock acquisition status
static LOCK_ACQUIRED: AtomicBool = AtomicBool::new(false);

/// Check if a process with given PID is still running.
/// Uses `kill -0` on Unix (works on both Linux and macOS, unlike /proc)
/// and `tasklist` on Windows.
#[cfg(unix)]
pub fn is_process_running(pid: u32) -> bool {
    std::process::Command::new("kill")
        .args(["-0", &pid.to_string()])
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status()
        .map(|status| status.success())
        .unwrap_or(false)
}

#[cfg(windows)]
pub fn is_process_running(pid: u32) -> bool {
    std::process::Command::new("tasklist")
        .args(["/FI", &format!("PID eq {}", pid), "/NH"])
        .output()
        .map(|output| {
            let stdout = String::from_utf8_lossy(&output.stdout);
            stdout.contains(&pid.to_string())
        })
        .unwrap_or(false)
}

#[cfg(not(any(unix, windows)))]
pub fn is_process_running(_pid: u32) -> bool {
    // On other platforms, conservatively assume process is not running
    false
}

/// Try to acquire an exclusive lock for cache extraction
/// Returns true if lock was acquired, false if cache is already being extracted
pub fn try_acquire_lock(paths: &WorkenvPaths) -> Result<bool> {
    // Create instance/extract directory if it doesn't exist
    let extract_dir = paths.extract();
    if let Err(e) = fs::create_dir_all(&extract_dir) {
        debug!("Failed to create extract directory: {}", e);
    }

    let lock_path = paths.lock_file();
    let pid = std::process::id();

    // Check for stale lock first
    if lock_path.exists() {
        debug!("🔍 Lock file exists, checking if it's stale...");

        // Try to read the PID from the lock file
        if let Ok(contents) = fs::read_to_string(&lock_path) {
            if let Ok(old_pid) = contents.trim().parse::<u32>() {
                if is_process_running(old_pid) {
                    debug!("🔒 Lock held by active process (PID: {old_pid})");
                    return Ok(false);
                } else {
                    info!("🧹 Removing stale lock from dead process (PID: {old_pid})");
                    fs::remove_file(&lock_path)?;
                }
            } else {
                // Invalid PID in lock file, remove it
                info!("🧹 Removing invalid lock file (couldn't parse PID)");
                fs::remove_file(&lock_path)?;
            }
        } else {
            // Can't read lock file, try to remove it
            info!("🧹 Removing unreadable lock file");
            fs::remove_file(&lock_path)?;
        }
    }

    // Try to create lock file exclusively
    match fs::OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(&lock_path)
    {
        Ok(mut file) => {
            // Write our PID to the lock file
            writeln!(file, "{pid}")?;
            debug!("🔒 Acquired extraction lock (PID: {pid})");
            LOCK_ACQUIRED.store(true, Ordering::SeqCst);
            Ok(true)
        }
        Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => {
            debug!("🔒 Lock file exists, another process is extracting");
            Ok(false)
        }
        Err(e) => Err(e.into()),
    }
}

/// Release the extraction lock
pub fn release_lock(paths: &WorkenvPaths) {
    let lock_path = paths.lock_file();
    if let Err(e) = fs::remove_file(&lock_path) {
        debug!("⚠️ Failed to remove lock file: {e}");
    } else {
        debug!("🔓 Released extraction lock");
    }
    LOCK_ACQUIRED.store(false, Ordering::SeqCst);
}

/// Wait for another process to finish extraction
pub fn wait_for_extraction(paths: &WorkenvPaths, timeout_secs: u64) -> Result<()> {
    use std::thread;
    use std::time::Duration;

    let lock_path = paths.lock_file();
    if !lock_path.exists() {
        debug!("✅ Extraction lock released, cache should be ready");
        thread::sleep(Duration::from_millis(100));
        return Ok(());
    }

    let max_attempts = timeout_secs * 10; // Check every 100ms

    for attempt in 0..max_attempts {
        if !lock_path.exists() {
            debug!("✅ Extraction lock released, cache should be ready");
            // Give a bit more time for files to be fully written
            thread::sleep(Duration::from_millis(100));
            return Ok(());
        }

        if attempt % 10 == 0 {
            debug!(
                "⏳ Waiting for extraction to complete... ({}/{}s)",
                attempt / 10,
                timeout_secs
            );
        }

        thread::sleep(Duration::from_millis(100));
    }

    Err(FlavorError::Generic(
        "Timeout waiting for cache extraction to complete".to_string(),
    ))
}

/// Mark cache extraction as complete
pub fn mark_extraction_complete(paths: &WorkenvPaths) -> Result<()> {
    let extract_dir = paths.extract();
    fs::create_dir_all(&extract_dir)?;
    let marker_path = paths.complete_file();
    let mut file = fs::File::create(&marker_path)?;
    writeln!(file, "{}", std::process::id())?;
    debug!("✅ Marked extraction as complete");
    Ok(())
}

/// Check if cache extraction is complete
pub fn is_extraction_complete(paths: &WorkenvPaths) -> bool {
    paths.complete_file().exists()
}

/// Mark cache as incomplete (used during signal handling)
pub fn mark_extraction_incomplete(paths: &WorkenvPaths) {
    let extract_dir = paths.extract();
    let _ = fs::create_dir_all(&extract_dir);
    // Note: We don't have an INCOMPLETE_FILE constant in the metadata architecture
    // This function might not be needed with atomic operations
    debug!("⚠️ Marked extraction as incomplete");
    // Remove the complete marker if it exists
    let _ = fs::remove_file(paths.complete_file());
}

/// Check if lock is currently acquired
pub fn is_lock_acquired() -> bool {
    LOCK_ACQUIRED.load(Ordering::SeqCst)
}

/// Clean up stale extraction directories from dead processes
pub fn cleanup_stale_extractions(paths: &WorkenvPaths) -> Result<()> {
    let tmp_dir = paths.tmp();

    // If the directory doesn't exist, nothing to clean
    if !tmp_dir.exists() {
        return Ok(());
    }

    // List all directories in tmp/
    if let Ok(entries) = fs::read_dir(&tmp_dir) {
        for entry in entries.flatten() {
            if let Ok(file_name) = entry.file_name().into_string() {
                // Try to parse PID from directory name
                if let Ok(pid) = file_name.parse::<u32>() {
                    // Check if process is still running
                    if !is_process_running(pid) {
                        let stale_dir = entry.path();
                        info!(
                            "🧹 Cleaning up stale extraction directory from dead process (PID: {})",
                            pid
                        );
                        if let Err(e) = fs::remove_dir_all(&stale_dir) {
                            debug!("⚠️ Failed to remove stale directory {:?}: {}", stale_dir, e);
                        }
                    }
                }
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;
    use tempfile::tempdir;

    #[test]
    fn wait_for_extraction_returns_ok_when_no_lock_exists_even_with_zero_timeout() {
        let temp = tempdir().expect("tempdir");
        let package = PathBuf::from("bundle.psp");
        let paths = WorkenvPaths::new(temp.path().to_path_buf(), package.as_path());

        assert!(wait_for_extraction(&paths, 0).is_ok());
    }

    #[test]
    fn try_acquire_lock_handles_invalid_lock_content() {
        let temp = tempdir().expect("tempdir");
        let package = PathBuf::from("bundle.psp");
        let paths = WorkenvPaths::new(temp.path().to_path_buf(), package.as_path());
        let lock_path = paths.lock_file();
        fs::create_dir_all(lock_path.parent().expect("lock parent")).expect("create lock parent");
        fs::write(&lock_path, "not-a-pid\n").expect("write invalid lock");

        assert!(try_acquire_lock(&paths).expect("lock acquisition"));
        assert!(is_lock_acquired());
        release_lock(&paths);
        assert!(!is_lock_acquired());
    }

    #[test]
    fn mark_extraction_complete_round_trips_marker_state() {
        let temp = tempdir().expect("tempdir");
        let package = PathBuf::from("bundle.psp");
        let paths = WorkenvPaths::new(temp.path().to_path_buf(), package.as_path());

        assert!(!is_extraction_complete(&paths));
        mark_extraction_complete(&paths).expect("mark complete");
        assert!(is_extraction_complete(&paths));
        mark_extraction_incomplete(&paths);
        assert!(!is_extraction_complete(&paths));
    }

    #[test]
    fn cleanup_stale_extractions_removes_dead_pid_directories() {
        let temp = tempdir().expect("tempdir");
        let package = PathBuf::from("bundle.psp");
        let paths = WorkenvPaths::new(temp.path().to_path_buf(), package.as_path());
        let stale_dir = paths.tmp().join("4294967295");
        let active_dir = paths.tmp().join("not-a-pid");

        fs::create_dir_all(&stale_dir).expect("create stale dir");
        fs::create_dir_all(&active_dir).expect("create active dir");

        cleanup_stale_extractions(&paths).expect("cleanup stale extractions");

        assert!(!stale_dir.exists());
        assert!(active_dir.exists());
    }
}
